include/nbl/builtin/hlsl/bxdf/ndf/microfacet_to_light_transform.hlsl

```diff
@@ -23,181 +23,50 @@ enum MicrofacetTransformTypes : uint16_t
+   {
+      SDualMeasureQuant<T> retval;
+      retval.microfacetMeasure = microfacetMeasure;
+      // do constexpr booleans first so optimizer picks up this and short circuits
+      const bool transmitted = reflect_refract==MTT_REFRACT || (reflect_refract!=MTT_REFLECT && VdotHLdotH < scalar_type(0.0));
+      retval.projectedLightMeasure = microfacetMeasure*mix<scalar_type>(scalar_type(0.25),VdotHLdotH,transmitted);
+      scalar_type denominator = clampedNdotV;
+      if (transmitted) // VdotHLdotH is negative under transmission, so thats denominator is negative
             denominator *= -VdotH_etaLdotH * VdotH_etaLdotH;
-        }
-        return pdf * (transmitted ? VdotHLdotH : scalar_type(0.25)) / denominator;
-    }
-
-    scalar_type pdf;
-    scalar_type absNdotV;
-    bool transmitted;
-    scalar_type VdotH;
-    scalar_type LdotH;
-    scalar_type VdotHLdotH;
-    scalar_type orientedEta;
+      retval.projectedLightMeasure /= denominator;
+      return retval;
+   }
 };
+}
 
 template<typename T>
-struct SDualMeasureQuant<T, true, MTT_REFLECT_REFRACT>
+SDualMeasureQuant<T> createDualMeasureQuantity(const T specialMeasure, typename vector_traits<T>::scalar_type clampedNdotV, typename vector_traits<T>::scalar_type clampedNdotL)
 {
-    using this_t = SDualMeasureQuant<T, true, MTT_REFLECT_REFRACT>;
-    using scalar_type = T;
-
-    NBL_CONSTEXPR_STATIC_INLINE MicrofacetTransformTypes Type = MTT_REFLECT_REFRACT;
-
-    scalar_type getMicrofacetMeasure()
-    {
-        return pdf;
-    }
-
-    scalar_type getProjectedLightMeasure()
-    {
-        scalar_type denominator = absNdotL;
-        if (transmitted)
-        {
-            const scalar_type VdotH_etaLdotH = (VdotH + orientedEta * LdotH);
-            // VdotHLdotH is negative under transmission, so thats denominator is negative
-            denominator *= -scalar_type(4.0) * VdotHLdotH / (VdotH_etaLdotH * VdotH_etaLdotH);
-        }
-        return pdf * denominator;
-    }
-
-    scalar_type pdf;
-    scalar_type absNdotL;
-    bool transmitted;
-    scalar_type VdotH;
-    scalar_type LdotH;
-    scalar_type VdotHLdotH;
-    scalar_type orientedEta;
-};
-
+   typename vector_traits<T>::scalar_type dummy;
+   return impl::createDualMeasureQuantity_helper<T,MTT_REFLECT>::__call(specialMeasure,clampedNdotV,clampedNdotL,dummy,dummy);
+}
+template<typename T, MicrofacetTransformTypes reflect_refract>
+SDualMeasureQuant<T> createDualMeasureQuantity(const T specialMeasure, typename vector_traits<T>::scalar_type clampedNdotV, typename vector_traits<T>::scalar_type clampedNdotL, typename vector_traits<T>::scalar_type VdotHLdotH, typename vector_traits<T>::scalar_type VdotH_etaLdotH)
+{
+   return impl::createDualMeasureQuantity_helper<T,reflect_refract>::__call(specialMeasure,clampedNdotV,clampedNdotL,VdotHLdotH,VdotH_etaLdotH);
```

Comment:
leave a little comment that the `specialMeasure` meands the measure definted by the specialization of `createDualMeasureQuantity_helper` and give an example of GGX as one thing that redefines it